calendar/page.js

```diff
@@ -304,31 +315,50 @@ class CalendarHandler {
     }
   }
 
-  // update calendar events based on the collection of records from the grist table.
-  async updateCalendarEvents(calendarEvents) {
-    // we need to keep track the ids of the events that are currently in the calendar to compare it
-    // with the new set of events when they come.
-    const currentIds = new Set();
-    for (const record of calendarEvents) {
-      // check if an event already exists in the calendar - update it if so, create new otherwise
-      const event = this.calendar.getEvent(record.id, CALENDAR_NAME);
-      const eventData = record;
-      if (!event) {
-        this.calendar.createEvents([eventData]);
+  getEvents() {
+    return this._allEvents;
+  }
+
+  setEvents(events) {
+    this._allEvents = events;
+  }
+
+  /**
+   * Adds/updates events that fall within the current date range, and removes
+   * events that do not.
+   */
+  renderVisibleEvents() {
+    const newVisibleEventIds = new Set();
+    const dateRangeStart = this.calendar.getDateRangeStart();
+    const dateRangeEnd = this.calendar.getDateRangeEnd().setHours(23, 99, 99, 999);
+
+    // Add or update events that are now visible.
+    for (const event of this._allEvents.values()) {
+      const isEventInRange = (
+        (event.start >= dateRangeStart && event.start <= dateRangeEnd) ||
+        (event.end >= dateRangeStart && event.end <= dateRangeEnd) ||
+        (event.start >= dateRangeStart && event.start <= dateRangeEnd) ||
```

Comment:
hang on what is this third one?

Reply:
Looks like a duplicate of the first. Thanks for catching that.